### PR TITLE
Alt/Plus

### PR DESCRIPTION
**Pull request by stephencelis**
Just some `Alt` and `Plus` stuff. Things fall down sometimes due to lack of conditional conformance, and it's too bad we can't get full `Alternative` support due to our inability to support `Applicative` _via_ higher kinded types.

### CODE REVIEW

File: Sources/Prelude/Optional.swift [hunk 1]
```diff
@@ -74,8 +74,23 @@ public func flatMap<A, B>(_ a2b: @escaping (A) -> B?) -> (A?) -> B? {
 // MARK: - Semigroup
 
 extension Optional where Wrapped: Semigroup {
-  public static func <> (lhs: Optional, rhs: Optional) -> Optional {
-    return curry(<>) <¢> lhs <*> rhs
```
Comment (stephencelis) [inline Sources/Prelude/Optional.swift]: Was reading this and realizing it didn't make sense! I'd done a copy-paste from `Either` without really thinking about it.

File: Sources/Prelude/Plus.swift [hunk 2]
```diff
@@ -0,0 +1,11 @@
+public protocol Plus: Alt {
+  static var empty: Self { get }
+}
+
+extension Array: Plus {}
```
Comment (stephencelis) [inline Sources/Prelude/Plus.swift]: Array gets `Plus` "for free" due do `Monoid` not being further constrained. It works here but could be a overloaded gotcha in the future. Haskell/PureScript uses `mempty`, Idris uses `neutral`, and I know you're a fan of `e`.

File: Sources/Prelude/Optional.swift (hunk 1 above)
Comment (mbrandonw) [inline Sources/Prelude/Optional.swift]: is the `switch` down below not equivalent to this definition? seems like it is to me

File: Sources/ValidationNearSemiring/ValidationNearSemiring.swift [hunk 3]
```diff
@@ -112,8 +112,8 @@ public func pure<E, A>(_ a: A) -> Validation<E, A> {
 
 // MARK: - Alt
 
-public extension Validation where E: NearSemiring {
-  public static func <|> (lhs: Validation, rhs: Validation) -> Validation {
+public extension Validation /* : Alt */ where E: NearSemiring {
```
Comment (mbrandonw) [inline Sources/ValidationNearSemiring/ValidationNearSemiring.swift]: we should come up with some kind of standard to mark future conditional conformances so it'll be easier to find use cases in the future...

File: Sources/Prelude/Optional.swift (hunk 1 above)
Comment (stephencelis) [inline Sources/Prelude/Optional.swift]: The `switch` does the opposite if `none` is around.

``` swift
// before
.some("hello") <> .none // .none
// after
.some("hello") <> .none // .some("hello")
```

**Comment by mbrandonw** [comment]
Oh yes yes I see!

**Pull request merged**